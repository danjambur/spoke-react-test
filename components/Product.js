import { gql, useQuery } from "@apollo/client";

export const GET_PRODUCT = gql`
  {
    product(id: $id) {
      title
      description
      onlineStoreUrl
    }
  }
`;

export default function Product({ product, id }) {
  console.log(id);
  const { loading, error, data } = useQuery(GET_PRODUCT, {
    variables: { id: "" },
  });

  console.log({ data });

  if (!product) return null;

  const price = product.priceRange.maxVariantPrice.amount;
  const hasImage = product.images.edges.length > 0;
  const lastUpdated = new Date(product.updatedAt);
  let threeDaysAgo = new Date();
  threeDaysAgo.setDate(threeDaysAgo.getDate() - 14);

  // set the URL for the image
  const imageSrc = hasImage
    ? product.images.edges[0].node.transformedSrc
    : null;

  // if its been updated in the last 3 days, i guess its new!
  let isNew = false;
  if (lastUpdated < threeDaysAgo) {
    isNew = true;
  }

  function goToProduct() {
    console.log("hello");
  }

  return (
    <div className="product" onClick={goToProduct}>
      {hasImage && <img className="product__image" src={imageSrc} />}
      <div className="product__desc">
        {<div>{isNew && <div>NEW</div>}</div>}
        <h4>{product.productType}</h4>
        <h2>
          {product.title} - Â£{price}
        </h2>
        <p>description goes here</p>
      </div>
    </div>
  );
}
