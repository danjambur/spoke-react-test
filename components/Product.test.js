import React from "react";
import Product from "./Product";
import renderer from "react-test-renderer";
import { MockedProvider } from "@apollo/client/testing";
import { GET_PRODUCT } from "./Product";

const productId = "Z2lkOi8vc2hvcGlmeS9Qcm9kdWN0LzE3MDU0MjY1ODM2MjE=";

const mock = [
  {
    request: {
      query: GET_PRODUCT,
      variables: {
        id: productId,
      },
    },
    result: {
      data: {
        node: {
          id: productId,
          images: {
            edges: [
              {
                node: {
                  transformedSrc:
                    "https://cdn.shopify.com/s/files/1/0057/7260/7557/products/1.Heroes-Jade_Flat.jpg?v=1573339341",
                },
              },
            ],
          },
          priceRange: {
            maxVariantPrice: {
              amount: "44.0",
            },
          },
          productType: "Heroes",
          tags: [
            "fabric: 50% off",
            "feed",
            "hidden sale item",
            "sale",
          ],
          title: "Mint",
        },
      },
    },
  },
];

it("should render loading state initially", () => {
  const component = renderer.create(
    <MockedProvider mocks={[]}>
      <Product />
    </MockedProvider>,
  );

  const tree = component.toJSON();
  expect(tree.children).toContain("Loading product...");
});

it("should render fine", () => {
  const component = renderer.create(
    <MockedProvider mocks={mock} addTypename={false}>
      <Product id={productId} />
    </MockedProvider>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

it("should render Product", async () => {
  const component = renderer.create(
    <MockedProvider mock={[mock]} addTypename={false}>
      <Product id={productId} />
    </MockedProvider>,
  );

  await new Promise((resolve) => setTimeout(resolve, 0)); // wait for response

  const h4 = component.root.findByType("h4");
  expect(h4.children).toContain("Heroes");
});
