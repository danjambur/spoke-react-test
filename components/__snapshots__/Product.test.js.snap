// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render fine 1`] = `
<p>
  Loading product...
</p>
`;
